qadence/analog/addressing.py

```diff
@@ -0,0 +1,163 @@
+from __future__ import annotations
+
+from dataclasses import dataclass
+from warnings import warn
+
+import sympy
+import torch
+from numpy import pi
+
+from qadence.parameters import Parameter, evaluate
+from qadence.types import StrEnum
+
+GLOBAL_MAX_AMPLITUDE = 300
+GLOBAL_MAX_DETUNING = 2 * pi * 2000
+LOCAL_MAX_AMPLITUDE = 3
+LOCAL_MAX_DETUNING = 2 * pi * 20
+
+
+class WeightConstraint(StrEnum):
+    """Supported types of constraints for addressing weights."""
+
+    NORMALIZE = "normalize"
+    """Normalize weights so that they sum up to 1."""
+
+    RESTRICT = "restrict"
+    """Restrict weight values to interval [0, 1]."""
+
+
+def sigmoid(x: torch.Tensor, a: float, b: float) -> sympy.Expr:
+    return 1.0 / (1.0 + sympy.exp(-a * (x + b)))
+
+
+@dataclass
+class AddressingPattern:
+    # number of qubits
+    n_qubits: int
+
+    # list of weights for fixed amplitude pattern that cannot be changed during the execution
+    weights_amp: dict[int, float | torch.Tensor | Parameter]
+
+    # list of weights for fixed detuning pattern that cannot be changed during the execution
+    weights_det: dict[int, float | torch.Tensor | Parameter]
+
+    # amplitude can also be chosen as a variational parameter if needed
+    amp: float | torch.Tensor | Parameter = LOCAL_MAX_AMPLITUDE
+
+    # detuning can also be chosen as a variational parameter if needed
+    det: float | torch.Tensor | Parameter = LOCAL_MAX_DETUNING
+
+    def _validate_weights(
+        self,
+        weights: dict[int, float | torch.Tensor | Parameter],
+    ) -> None:
+        for v in weights.values():
+            if not isinstance(v, Parameter):
+                if not (v >= 0.0 and v <= 1.0):
+                    raise ValueError("Addressing pattern weights must sum fall in range [0.0, 1.0]")
+
+    def _constrain_weights(
+        self,
+        weights: dict[int, float | torch.Tensor | Parameter],
+    ) -> dict:
+        # augment weight dict if needed
+        weights = {
+            i: Parameter(0.0)
+            if i not in weights
+            else (Parameter(weights[i]) if not isinstance(weights[i], Parameter) else weights[i])
+            for i in range(self.n_qubits)
+        }
+
+        # restrict weights to [0, 1] range
+        weights = {
+            k: abs(v * (sigmoid(v, 20, 1.0) - sigmoid(v, 20.0, -1.0))) for k, v in weights.items()
+        }
```

Comment:
If I recall correctly this makes it so that any weight higher than 1.0 or lower than -1.0 gets effectively counted as 0 right? Can you add a line to the comment saying what it does?

Reply:
expanded the comment